p2p/transport/webrtc/udp_mux_conn.go

```diff
@@ -0,0 +1,246 @@
+package libp2pwebrtc
+
+import (
+	"encoding/binary"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/pion/logging"
+	"github.com/pion/transport/packetio"
+)
+
+type udpMuxedConnParams struct {
+	Mux       *UDPMuxNewAddr
+	AddrPool  *sync.Pool
+	Key       string
+	LocalAddr net.Addr
+	Logger    logging.LeveledLogger
+}
+
+// udpMuxedConn represents a logical packet conn for a single remote as identified by ufrag
+type udpMuxedConn struct {
+	params *udpMuxedConnParams
+	// remote addresses that we have sent to on this conn
+	addresses []string
+
+	// channel holding incoming packets
+	buffer     *packetio.Buffer
+	closedChan chan struct{}
+	closeOnce  sync.Once
+	mu         sync.Mutex
+}
+
+func newUDPMuxedConn(params *udpMuxedConnParams) *udpMuxedConn {
+	p := &udpMuxedConn{
+		params:     params,
+		buffer:     packetio.NewBuffer(),
+		closedChan: make(chan struct{}),
+	}
+
+	return p
+}
+
+func (c *udpMuxedConn) ReadFrom(b []byte) (n int, raddr net.Addr, err error) {
+	buf := c.params.AddrPool.Get().(*bufferHolder)
+	defer c.params.AddrPool.Put(buf)
+
+	// read address
+	total, err := c.buffer.Read(buf.buffer)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	dataLen := int(binary.LittleEndian.Uint16(buf.buffer[:2]))
```

Comment:
This will panic if `Read` read less than 2 bytes.

Reply:
https://github.com/libp2p/go-libp2p/pull/1655/files#diff-f175016b12b6cf30881b97cb5812f76ce03a6ef6370ccf3be2664c9114b8b96dR181 the format should ensure we never write less than 2 bytes. I'm not confident about this though since they do not verify bytes written here: https://github.com/libp2p/go-libp2p/pull/1655/files#diff-f175016b12b6cf30881b97cb5812f76ce03a6ef6370ccf3be2664c9114b8b96dR203